src/backend/TrafficCourts/Common/Features/FilePersistence/MinioFilePersistenceService.cs

```diff
@@ -122,4 +125,128 @@ public override async Task<string> SaveFileAsync(MemoryStream data, Cancellation
+        {
+            ["FileName"] = filename,
+            ["BucketName"] = _bucketName,
+        });
+
+        try
+        {
+            // bucket must exist prior to calling this otherwise BucketNotFoundException will be thrown
+            // in the shared services object store, we are not giving permissions to list or create buckets
+
+            // Upload a file to bucket.
+            PutObjectArgs putObjectArgs = new PutObjectArgs()
+                    .WithBucket(_bucketName)
+                    .WithObject(filename)
+                    .WithContentType("application/json")
+                    .WithObjectSize(dataJsonStream.Length)
+                    .WithStreamData(dataJsonStream)
+                    .WithHeaders(headers);
+
+            await _objectOperations.PutObjectAsync(putObjectArgs, cancellationToken);
+            return filename;
+        }
+        catch (BucketNotFoundException exception)
+        {
+            _logger.LogWarning(exception, "Object Store bucket not found");
+            throw new FileNotFoundException("File not found", filename, exception);
+        }
+        catch (MinioException exception)
+        {
+            _logger.LogError(exception, "Failed to upload file to object storage");
+            throw new MinioFilePersistenceException("Failed to upload file to object storage", exception);
+        }
+    }
+
+    public override async Task<T?> GetJsonDataAsync<T>(string filename, CancellationToken cancellationToken) where T : class
+    {
+        ArgumentNullException.ThrowIfNull(filename);
+
+        Type targetType = typeof(T);
+
+        using var scope = _logger.BeginScope(new Dictionary<string, object>
+        {
+            ["FileName"] = filename,
+            ["BucketName"] = _bucketName,
+        });
+
+        try
+        {
+            MemoryStream objectStream = _memoryStreamManager.GetStream();
+
+            GetObjectArgs args = new GetObjectArgs()
+                .WithBucket(_bucketName)
+                .WithObject(filename)
+                .WithCallbackStream((stream) => { stream.CopyTo(objectStream); });
+
+            ObjectStat? status = await _objectOperations.GetObjectAsync(args, cancellationToken);
+            if (status == null)
+            {
+                _logger.LogInformation("Could not fetch object from object storage with objectStat and headers");
+                return null;
+            }
+            Dictionary<string, string> headers = status.MetaData;
+            if (headers.TryGetValue("type", out string? objectType))
+            {
+                if (objectType != targetType.Name)
+                {
+                    _logger.LogWarning("Target object type to return: {targetType} does not match the type of json object fetched from object storage: {objectType}", targetType.Name, objectType);
+                    return null;
+                }
+            }
+            string jsonString = Encoding.UTF8.GetString(objectStream.ToArray());
```

Comment:
Deserialize directly from the memory stream. You do not need to convert the memory stream to a byte array, then to a string and then deserialize.

Reply:
Done.